src/storage/src/sink/iceberg.rs

```diff
@@ -0,0 +1,1413 @@
+                                // Try writing the row to any in-flight batch it belongs to...
+                                for (batch_desc, delta_writer) in in_flight_batches.iter_mut() {
+                                    let (lower, upper) = batch_desc;
+                                    if PartialOrder::less_equal(lower, &ts_antichain)
+                                        && PartialOrder::less_than(&ts_antichain, upper)
+                                    {
+                                        let record_batch = row_to_recordbatch(
+                                            diff_pair.clone(),
+                                            Arc::clone(&arrow_schema),
+                                            Arc::clone(&schema_with_op),
+                                        )
+                                        .context("failed to convert row to recordbatch")?;
+                                        delta_writer
+                                            .write(record_batch)
+                                            .await
+                                            .context("Failed to write row to DeltaWriter")?;
+                                        written = true;
+                                        break;
+                                    }
+                                }
+                                if !written {
+                                    // ...otherwise stash it for later
+                                    let entry = stashed_rows.entry(row_ts).or_default();
+                                    entry.push((row, diff_pair));
+                                }
+                            }
+                        }
+                        Event::Progress(frontier) => {
+                            input_frontier = frontier;
+                        }
+                    }
+                }
+
+                // Check if frontiers have advanced, which may unlock batches ready to close
+                if PartialOrder::less_equal(
+                    &processed_batch_description_frontier,
+                    &batch_description_frontier,
+                ) || PartialOrder::less_than(&processed_input_frontier, &input_frontier)
+                {
+                    // Close batches whose upper is now in the past
+                    let mut ready_batches: Vec<_> = in_flight_batches
+                        .extract_if(|(lower, upper), _| {
+                            PartialOrder::less_than(lower, &batch_description_frontier)
+                                && PartialOrder::less_than(upper, &input_frontier)
+                        })
+                        .collect();
+
+                    // Sort by upper bound to process in timestamp order
+                    ready_batches.sort_by(|(a, _), (b, _)| {
+                        a.1.elements()
+                            .into_element()
+                            .cmp(b.1.elements().into_element())
```

Comment:
This will panic if the sinked collection reaches the empty frontier. You can sort frontiers like so: https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/render/persist_sink.rs#L1102-L1110